Update all GNU docker images to latest LTS version on Docker.

Alexhuszagh opened the pull request:
Increment Ubuntu base image versions to 20.04.

Update linux-image script to latest kernel and debian versions.

Update by default to kernel version 5.10.0-8. This means updating our debian source to bullseye from buster. 32-bit big-endian mips was discontinued in bullseye, so we revert to buster. For some images, due to constantly updating linux kernel versions, we need to use wildcards otherwise the build step breaks. Since there may be more than one relevant package, we've added a function to manually expand wildcards and select the best kernel version, `max_kernel_version`. Likewise, on 32-bit big-endian mips, we need to specify the ncurses version.

Created temporary symlinks for autconf and autom4te due to the build expecting a hard-coded version (2.69) of these binaries. Fixed the patch for `debian/rules` due to changed line numbers. Updated the patch to use dwarf rather than sjlj exceptions to patch the template file (`debian/gcc-mingw-w64-i686.install.in`) since `debian/gcc-mingw-w64-i686.install` is overwritten during the build.

For `x86_64-unknown-linux-gnu`, building the linux image fails unless we download specific versions of `libgcc-s1` and `libstdc++6`, since the pre-installed Ubuntu versions are higher than the Debian versions. We therefore extract the specific versions. However, while building the linux image, it prefers these system versions, so we must uninstall them or else while running `qemu-system` it cannot find `libgcc_s1.so.1`. Since `apt` and basically every other package besides `dpkg` relies on `libgcc-s1`, we have to temporarily delete it and reinstall it.

Closes #616.
Closes #557. We've already addressed the Qemu fixes, which will be applied automatically.
Closes #517.
Closes #417.

Replaces #481.

Alexhuszagh commented:
Most of the build errors seem to be from spurious network errors, but if any other of the build steps are failing, please let me know so I can modify the build steps.

deftsp commented:
```
./build-docker-image.sh x86_64-unknown-linux-gnu
```

```
#20 96.75 + apt-get update
#20 97.32 Get:1 http://security.debian.org bullseye-security InRelease [44.1 kB]
#20 97.83 Get:2 http://security.debian.org bullseye-security/main amd64 Packages [35.7 kB]
#20 103.1 Get:3 http://cdn-fastly.deb.debian.org/debian bullseye InRelease [113 kB]
#20 103.7 Get:4 http://cdn-fastly.deb.debian.org/debian bullseye/main amd64 Packages [11.1 MB]
#20 110.0 Fetched 11.3 MB in 13s (855 kB/s)
#20 110.0 Reading package lists...
#20 110.5 + mkdir -p /qemu/amd64
#20 110.5 + chmod 777 /qemu /qemu/amd64
#20 110.5 + cd /qemu/amd64
#20 110.5 + apt-get -d --no-install-recommends download busybox:amd64 dropbear-bin:amd64 libtommath1:amd64 libtomcrypt1:amd64 libgmp10:amd64 libc6:amd64 libgcc-s1:amd64 libstdc++6:amd64 linux-image-5.10.0-8-amd64:amd64 ncurses-base zlib1g:amd64
#20 111.3 E: Can't find a source to download version '10.3.0-1ubuntu1~20.04' of 'libgcc-s1:amd64'
#20 111.3 E: Can't find a source to download version '10.3.0-1ubuntu1~20.04' of 'libstdc++6:amd64'
#20 111.3 E: Can't find a source to download version '1:1.2.11.dfsg-2ubuntu1.2' of 'zlib1g:amd64'
```

ChriFo commented:
To build the `i686-pc-windows-gnu` image I had to adjust the  [`mingw.sh`](https://github.com/rust-embedded/cross/blob/master/docker/mingw.sh):

```diff
@@ -37,7 +37,7 @@
     pushd gcc-mingw-w64-*

     # We are using dwarf exceptions instead of sjlj
-    sed -i -e 's/libgcc_s_sjlj-1/libgcc_s_dw2-1/g' debian/gcc-mingw-w64-i686.install
+    sed -i -e 's/libgcc_s_sjlj-1/libgcc_s_dw2-1/g' debian/gcc-mingw-w64-i686.install.in

     # Only build i686 packages (disable x86_64)
     patch -p0 <<'EOF'
@@ -90,8 +90,7 @@
      BUILD_TARGET :=
      INSTALL_TARGET := install install-lto-plugin
  endif
-@@ -85,7 +85,7 @@
- 	sed -i 's/@@VERSION@@/$(target_version)/g' debian/control
+@@ -87,6 +87,6 @@
  	touch $@

 -targets := i686-w64-mingw32 x86_64-w64-mingw32
@@ -99,7 +98,7 @@
  threads := posix win32

  # Hardening on the host, none on the target
-@@ -216,6 +216,10 @@
+@@ -220,6 +220,10 @@
  # Enable libatomic
  CONFFLAGS += \
  	--enable-libatomic
@@ -107,9 +106,18 @@
 +CONFFLAGS += \
 +	--disable-sjlj-exceptions \
 +	--with-dwarf2
- # Enable experimental::filesystem
+ # Enable experimental::filesystem and std::filesystem
  CONFFLAGS += \
  	--enable-libstdcxx-filesystem-ts=yes
```

ChriFo commented:
> ```
> ./build-docker-image.sh x86_64-unknown-linux-gnu
> ```
> 
> ```
> #20 96.75 + apt-get update
> #20 97.32 Get:1 http://security.debian.org bullseye-security InRelease [44.1 kB]
> #20 97.83 Get:2 http://security.debian.org bullseye-security/main amd64 Packages [35.7 kB]
> #20 103.1 Get:3 http://cdn-fastly.deb.debian.org/debian bullseye InRelease [113 kB]
> #20 103.7 Get:4 http://cdn-fastly.deb.debian.org/debian bullseye/main amd64 Packages [11.1 MB]
> #20 110.0 Fetched 11.3 MB in 13s (855 kB/s)
> #20 110.0 Reading package lists...
> #20 110.5 + mkdir -p /qemu/amd64
> #20 110.5 + chmod 777 /qemu /qemu/amd64
> #20 110.5 + cd /qemu/amd64
> #20 110.5 + apt-get -d --no-install-recommends download busybox:amd64 dropbear-bin:amd64 libtommath1:amd64 libtomcrypt1:amd64 libgmp10:amd64 libc6:amd64 libgcc-s1:amd64 libstdc++6:amd64 linux-image-5.10.0-8-amd64:amd64 ncurses-base zlib1g:amd64
> #20 111.3 E: Can't find a source to download version '10.3.0-1ubuntu1~20.04' of 'libgcc-s1:amd64'
> #20 111.3 E: Can't find a source to download version '10.3.0-1ubuntu1~20.04' of 'libstdc++6:amd64'
> #20 111.3 E: Can't find a source to download version '1:1.2.11.dfsg-2ubuntu1.2' of 'zlib1g:amd64'
> ```
Maybe a bad hack... but when adding the following lines under line 12 in [`linux-image.sh`](https://github.com/rust-embedded/cross/blob/master/docker/linux-image.sh):
```
    debsource="${debsource}\ndeb http://de.archive.ubuntu.com/ubuntu focal main"
    debsource="${debsource}\ndeb http://de.archive.ubuntu.com/ubuntu focal-updates main"
```
building `x86_64-unknown-linux-gnu` succeeded.

Xuanwo commented:
What's going on here? Is there any help I can provide?

Amanieu commented:
Any updates on this?

svenstaro commented:
> Any updates on this?

Currently needs someone to fix https://github.com/cross-rs/cross/pull/609. Perhaps you can help?

Emilgardis commented:
bors try --target unknown-linux

Alexhuszagh commented:
Sorry this should be converted to a draft: this commit was simply to rebase to the current codebase, to avoid merge conflicts, and I'm tracking the list of all images that fail to build and annotating the reasons for their failure.

Alexhuszagh commented:
> What's going on here? Is there any help I can provide?

@Xuanwo Sorry for the delay, I'm currently fixing the following images:

- mips-unknown-linux-gnu
    - `/usr/bin/ld: cannot find -lmount`
- mipsel-unknown-linux-gnu
    - `Unable to locate package linux-image-5.10.0-8-4kc-malta:mipsel`
- mips64el-unknown-linux-gnuabi64
    - `Unable to locate package linux-image-5.10.0-8-5kc-malta:mips64el`
- powerpc64le-unknown-linux-gnu
    - `Unable to locate package linux-image-5.10.0-8-powerpc64le:ppc64el`
- s390x-unknown-linux-gnu
    - ` Unable to locate package linux-image-5.10.0-8-s390x:s390x`
- x86_64-unknown-linux-gnu
    - `Can't find a source to download version '10.3.0-1ubuntu1~20.04' of 'libgcc-s1:amd64'`

This is non-exhaustive and some of these should be trivial to fix, but if you are willing to submit PRs for any of these, it would be great.

For some, like `libgcc-s1`, it's a simple issue of the package name changing. For some, the supported kernel versions have changed, and for a few this requires essentially building everything from source if I remember correctly.

If you would like to submit patches for any of these, feel free to submit a PR on my fork, and I'd recommend starting at their of the first 2.

Make sure you run the following commands to test the code:

```bash
# This is an example export, all can be seen in `build_all.sh`.
# export TARGET=arm-unknown-linux-musleabihf STD=1 RUN=1 
./build-docker-image.sh "$TARGET"
ci/test.sh
```

This should require a nightly compiler for the tests, and for some very recent ones, I've been getting an issue with the manifest being unknown after some tests run. This isn't a big issue and they pass the CI tests, so failing with that error counts as a successful build locally.

Alexhuszagh commented:
In addition, the package `armv5te-unknown-linux-gnueabi` fails because of https://github.com/cross-rs/cross/pull/593. This is because `tzdata` requires interactive prompts, so it cannot be built in an automated manner. Merging #593 should fix this issue.

Alexhuszagh commented:
Ok I've added numerous fixes for updating the Docker Ubuntu versions, and the only issues remaining are:

**Interactive Issues**

- armv5te-unknown-linux-gnueabi
- i686-pc-windows-gnu

**Package Name Issues**

- x86_64-unknown-linux-gnu: `Can't find a source to download version '10.3.0-1ubuntu1~20.04' of 'libgcc-s1:amd64'`

**Docker Dependency Issue**

- asmjs-unknown-emscripten:  `failed to load cache key: docker.io/trzeci/emscripten:1.39.20-upstream: not found`
- wasm32-unknown-emscripten: `failed to load cache key: docker.io/trzeci/emscripten:1.39.20-upstream: not found`

**Other Issues**

- x86_64-unknown-linux-gnu: `Can't find a source to download version '10.3.0-1ubuntu1~20.04' of 'libgcc-s1:amd64'`
- x86_64-apple-darwin (likely due to missing build dependencies since I'm running Linux)
- x86_64-pc-windows-msvc (likely due to missing build dependencies since I'm running Linux)

**Kernel Selection Logic**

@Emilgardis Due to the use of backports or certain kernel versions which might not be stable (for example, `5.10.0-8` isn't present anymore on `s390x`, however, `5.10.0-9` is, I've added the following code to select the best (newest) kernel version from a wildcard:

```bash
max_kernel_version() {
    # kernel versions have the following format:
    #   `5.10.0-10-$arch`, where the `$arch` may be optional.
    local IFS=$'\n'
    local -a versions
    local major=0
    local minor=0
    local patch=0
    local release=0
    local index=0

    read -r -d '' -a versions <<< "$1"
    for i in "${!versions[@]}"; do
        local version="${versions[$i]}"
        local x=$(echo "$version" | cut -d '.' -f 1)
        local y=$(echo "$version" | cut -d '.' -f 2)
        local z=$(echo "$version" | cut -d '.' -f 3 | cut -d '-' -f 1)
        local r=$(echo "$version" | cut -d '-' -f 2)
        local is_larger=

        if [ "$x" -gt "$major" ]; then
            is_larger=1
        elif [ "$x" -eq "$major" ] && [ "$y" -gt "$minor" ]; then
            is_larger=1
        elif [ "$x" -eq "$major" ] && [ "$y" -eq "$minor" ] && [ "$z" -gt "$patch" ]; then
            is_larger=1
        elif [ "$x" -eq "$major" ] && [ "$y" -eq "$minor" ] && [ "$z" -eq "$patch" ] && [ "$r" -gt "$release" ]; then
            is_larger=1
        fi

        if [ -n "$is_larger" ]; then
            index="$i"
            major="$x"
            minor="$y"
            patch="$z"
            release="$r"
        fi
    done

    echo "${versions[index]}"
}
```

Any review of this code would be wonderful.

Emilgardis commented:
> * asmjs-unknown-emscripten:  `failed to load cache key: docker.io/trzeci/emscripten:1.39.20-upstream: not found`
> 
> * wasm32-unknown-emscripten: `failed to load cache key: docker.io/trzeci/emscripten:1.39.20-upstream: not found`

Feel free to ignore these, they are disabled in CI

Emilgardis commented:
@Alexhuszagh I added you as a member to bors so you can run `bors try`

please fix the conflict

Alexhuszagh commented:
@Emilgardis The only issue is the CHANGELOG correct?

Alexhuszagh commented:
bors+

Alexhuszagh commented:
bors try

Alexhuszagh commented:
Ok the 2 remaining failing images:
- armv5te-unknown-linux-gnueabi: caused by the interactive build, fixed in my other PR
- sparc64-unknown-linux-gnu (currently debugging).

Alexhuszagh commented:
@Emilgardis 

Ok it seems the latest sparc64 Linux kernel image doesn't have any real files in it, which wasn't true a little while ago. Using an older package ( `linux-image-5.16.0-6:sparc64`) fixes this.

```
# dpkg-deb -c sparc64/linux-image-sparc64_5.16.18-1_sparc64.deb 
drwxr-xr-x root/root         0 2022-03-29 20:46 ./
drwxr-xr-x root/root         0 2022-03-29 20:46 ./usr/
drwxr-xr-x root/root         0 2022-03-29 20:46 ./usr/share/
drwxr-xr-x root/root         0 2022-03-29 20:46 ./usr/share/bug/
drwxr-xr-x root/root         0 2022-03-29 20:46 ./usr/share/bug/linux-image-sparc64/
-rw-r--r-- root/root       239 2022-03-29 20:46 ./usr/share/bug/linux-image-sparc64/presubj
drwxr-xr-x root/root         0 2022-03-29 20:46 ./usr/share/doc/
lrwxrwxrwx root/root         0 2022-03-29 20:46 ./usr/share/doc/linux-image-sparc64 -> linux-image-5.16.0-6-sparc64
```

This is very different, say, the `s390x`, which has the following package [contents](https://packages.debian.org/stretch-backports/s390x/linux-image-4.19.0-0.bpo.9-s390x/filelist). If we use the package `linux-image-5.16.0-6:sparc64`, which isn't the latest version, it works. So we probably need to hardcode the kernel version here.

This is a separate issue, but I'm more than happy to fix it here.

Alexhuszagh commented:
bors try

Emilgardis commented:
Wierd!

https://packages.debian.org/sid/linux-image-5.16.0-6-sparc64

It technically is 5.16.18 I guess

Emilgardis commented:
seems like this is all done then, that try seems to be just a random failure.

Alexhuszagh commented:
@Emilgardis I've had to fix a few of the patches for i686-pc-windows-gnu, and I've got one minor question. The dpkg build script wants a hard-coded version of autoconf, which setting the `AUTOCONF` environment variable to the system default (`/usr/bin/autoconf`) doesn't seem to fix.

The following, however, does work:

```bash
main() {
    ...

    # Need symlinks for specific autoconf versions, since it
    # attempts to use autoconf2.69 and autom4te2.69.
    ln -s /usr/bin/autoconf /usr/bin/autoconf2.69
    ln -s /usr/bin/autom4te /usr/bin/autom4te2.69
    
    ...

    # Unlink our temporary aliases
    unlink /usr/bin/autoconf2.69
    unlink /usr/bin/autom4te2.69
}
```

This seems very fragile, but I'm not sure another solution.

Emilgardis commented:
I'm not sure I follow, can you push these new patches for gcc on mingw to make it clearer what the issue is

Alexhuszagh commented:
I'm actually kind of confused. Um, I'll be working on this for a bit, I'll get back before the end of the weekend if I can.

### Issue

**autoconf**

If I run `bash mingw.sh`, it fails during the `MAKEFLAGS=--silent dpkg-buildpackage -nc -B --jobs=auto` step, with the following error:

```bash
Running autoconf2.69 in src/...
Running autoconf2.69 in src/gcc/...
sh: 3: autoconf2.69: not found
Running autoconf2.69 in src/libcc1/...
sh: 3: autoconf2.69: not found
sh: 3: autoconf2.69: not foundRunning autoconf2.69 in src/libdecnumber/...
```

Trying to manually specify the autoconf binary via the `AUTOCONF` environment variable doesn't work, in either of the cases below (I'll retry, just in case):

```bash
$ MAKEFLAGS=--silent AUTOCONF=/usr/bin/autoconf dpkg-buildpackage -nc -B --jobs=auto
$ export AUTOCONF=/usr/bin/autoconf
$ MAKEFLAGS=--silent dpkg-buildpackage -nc -B --jobs=auto
```

The following, however, does work:

```bash
ln -s /usr/bin/autoconf /usr/bin/autoconf2.69
ln -s /usr/bin/autom4te /usr/bin/autom4te2.69
```

The documentation for `src/libphobos/configure.ac`, which was the only file that manually specified `autoconf2.69` via a recursive grep, says:

```Makefile
# This requires that you have your environment set-up to use explicit
# versions of automake and autoconf.
#
#    export ACLOCAL=/usr/bin/aclocal-1.15
#    export AUTOMAKE=/usr/bin/automake-1.15
#    export AUTOM4TE=/usr/bin/autom4te2.69
#    export AUTOCONF=/usr/bin/autoconf2.69
#
#    autoreconf2.69
```

**Missing Files**

If that's patched, we get an error of the following:

```bash
warning: gcc-mingw-w64-i686 missing files: usr/lib/gcc/i686-w64-mingw32/9*/libgcc_s_sjlj-1.dll
```

This is unusual, since a simple `ls` shows 2 satisfactory files (unless the issue is it using a local, rather than a global path):

```bash
$ ls usr/lib/gcc/i686-w64-mingw32/9*/libgcc_s_sjlj-1.dll
usr/lib/gcc/i686-w64-mingw32/9.3-posix/libgcc_s_sjlj-1.dll  usr/lib/gcc/i686-w64-mingw32/9.3-win32/libgcc_s_sjlj-1.dll
```

### Other Fixes

There was a minor change in the `mingw.sh` patch, which was due to changing line numbers.

```diff
diff --git a/docker/mingw.sh b/docker/mingw.sh
index 196a874..2573233 100755
--- a/docker/mingw.sh
+++ b/docker/mingw.sh
@@ -99,7 +99,7 @@ EOF
  threads := posix win32
 
  # Hardening on the host, none on the target
-@@ -216,6 +216,10 @@
+@@ -220,6 +220,10 @@
  # Enable libatomic
  CONFFLAGS += \
  	--enable-libatomic
@@ -107,11 +107,16 @@ EOF
 +CONFFLAGS += \
 +	--disable-sjlj-exceptions \
 +	--with-dwarf2
- # Enable experimental::filesystem
+ # Enable experimental::filesystem and std::filesystem
  CONFFLAGS += \
  	--enable-libstdcxx-filesystem-ts=yes
 EOF
```

Alexhuszagh commented:
bors try

Alexhuszagh commented:
Ok I'm converting this from a draft to be ready for review. The only issue with `i686-pc-windows-gnu` seems to be now a bors artifact (missing log), and the images build for me locally without issue. Any feedback would be appreciated.

Emilgardis commented:
bors try --target gnu

I want to make sure that the readme accurately reflects gcc and libc versions, does it?

Jules-Bertholet commented on docker/linux-image.sh:
```diff
@@ -58,27 +115,26 @@ main() {
             # https://packages.debian.org/en/sid/linux-image-sparc64
-            kernel='*-sparc64'
-            libgcc="libgcc-s1"
+            # the latest image doesn't work as of 5.16.18-1
+            kernel='5.16.0-6-sparc64'
```
```suggestion
            kernel='5.*-sparc64'
```
The latest image is now [15.17.`](https://packages.debian.org/sid/linux-image-5.17.0-1-sparc64), which has all the needed files. (This change should resolve the build failure)

Alexhuszagh replied:
Let me test, because a wildcard last time led to an empty package being used without the proper files even though I was worried this exact issue would happen.

Alexhuszagh replied:
Yeah 5.17.6 doesn't work: it's a dummy [package](https://packages.debian.org/en/sid/sparc64/linux-image-sparc64/filelist). We should be able to get away with [5.17.2](https://packages.debian.org/sid/sparc64/linux-image-5.17.0-2-sparc64/filelist) however, but this is a very temporary fix.

```text
/usr/share/bug/linux-image-sparc64/presubj
/usr/share/doc/linux-image-sparc64
```

Emilgardis commented:
lets see

bors try --target sparc64-unknown-linux-gnu

Jules-Bertholet commented:
@Emilgardis you need to apply the change before testing it…

Emilgardis commented:
> @Emilgardis you need to apply the change before testing it…

oh wow, I misread your suggested change as the actual existing change :D

Emilgardis commented:
seems fine to me, kinda wonky but seems this is how they are maintaining the package so we'll have to adapt

Alexhuszagh commented:
Ok I'm pretty sure the wildcard was a bug after looking it over carefully, and I've merged @Jules-Bertholet's commit, but I've added a working solution on another [branch](https://github.com/Alexhuszagh/cross/tree/dummy) for removing dummy packages if this ever becomes an issue in the future.

The logic is fairly wonky, because it turns out `apt-file` doesn't work with these mirrors, and the only solution is to download the package, query the contents, count the number of files, and then repeat this. Far from ideal, but I know of no other solutions. This is the function to filter out the packages, but I'm fairly reticent of committing this as-is.

```bash
remove_dummy_packages() {
    # sometimes the linux-image on sparc64 can have only documentation
    # which causes a failing install, so a lower version must be used.
    # this removes the dummy packages, forcing 50+ files in a package.
    # this might have been a once-off bug, so we might remove this later
    local IFS=$'\n'
    local -a packages
    local -a real_packages=()
    local package
    local file_list
    local files
    local count

    read -r -d '' -a packages <<< "$1"
    for package in "${packages[@]}"; do
        # the results have the format $package: $file, so
        # we want to count the number of lines
        apt-get -d --no-install-recommends download "$package" -qq 2>/dev/null >/dev/null
        file_list="$(dpkg --contents "$package"_*.deb)"
        read -r -d '' -a files <<< "$file_list"
        count="${#files[@]}"

        # the real packages have many > 1000 files, so a high threshold is fine
        if [ "$count" -gt 50 ]; then
            real_packages+=("$package")
        fi

        rm "$package"_*.deb
    done

    echo "${real_packages[*]}"
}
```

Alexhuszagh commented:
bors try

Alexhuszagh commented:
@Emilgardis Bors succeeded, so all checks seem to pass. Should I squash the commits, or should this be satisfactory?

Emilgardis commented:
Awesome!

I want to make sure the versions mentioned in the readme are correct, could you do that?

Emilgardis commented:
also, no need to squash unless you want to

Alexhuszagh commented:
> 



> Awesome!
> 
> I want to make sure the versions mentioned in the readme are correct, could you do that?

Working on this, I'm assuming this is for this section?

```markdown

| Target                               |  libc  |   GCC   | C++ | QEMU  | `test` |
|--------------------------------------|-------:|--------:|:---:|------:|:------:|
| `*-apple-ios` [1]                    | N/A    | N/A     | N/A | N/A   |   ✓    |
| `aarch64-linux-android` [2]          | 9.0.0  | 4.9     | ✓   | 5.1.0 |   ✓    |
| `aarch64-unknown-linux-gnu`          | 2.19   | 4.8.2   | ✓   | 5.1.0 |   ✓    |
| `aarch64-unknown-linux-musl`         | 1.1.20 | 6.3.0   |     | 5.1.0 |   ✓    |
| `arm-linux-androideabi` [2]          | 9.0.0  | 4.9     | ✓   | 5.1.0 |   ✓    |
| `arm-unknown-linux-gnueabi`          | 2.19   | 4.8.2   | ✓   | 5.1.0 |   ✓    |
| `arm-unknown-linux-gnueabihf`        | 2.27   | 7.3.0   | ✓   | 5.1.0 |   ✓    |
| `arm-unknown-linux-musleabi`         | 1.1.20 | 6.3.0   |     | 5.1.0 |   ✓    |
| `arm-unknown-linux-musleabihf`       | 1.1.20 | 6.3.0   |     | 5.1.0 |   ✓    |
| `armv5te-unknown-linux-gnueabi`      | 2.27   | 7.5.0   | ✓   | 5.1.0 |   ✓    |
| `armv5te-unknown-linux-musleabi`     | 1.1.20 | 6.3.0   |     | 5.1.0 |   ✓    |
| `armv7-linux-androideabi` [2]        | 9.0.0  | 4.9     | ✓   | 5.1.0 |   ✓    |
| `armv7-unknown-linux-gnueabihf`      | 2.15   | 4.6.2   | ✓   | 5.1.0 |   ✓    |
| `armv7-unknown-linux-musleabihf`     | 1.1.20 | 6.3.0   |     | 5.1.0 |   ✓    |
| `asmjs-unknown-emscripten` [6]       | 1.2.2  | 3.1.10  | ✓   | N/A   |   ✓    |
| `i586-unknown-linux-gnu`             | 2.23   | 5.3.1   | ✓   | N/A   |   ✓    |
| `i586-unknown-linux-musl`            | 1.1.20 | 6.3.0   |     | N/A   |   ✓    |
| `i686-unknown-freebsd` [4]           | 12.1   | 6.4.0   |     | N/A   |   ✓    |
| `i686-linux-android` [2]             | 9.0.0  | 4.9     | ✓   | 5.1.0 |   ✓    |
| `i686-pc-windows-gnu`                | N/A    | 7.3.0   | ✓   | N/A   |   ✓    |
| `i686-unknown-linux-gnu`             | 2.15   | 4.6.2   | ✓   | N/A   |   ✓    |
| `i686-unknown-linux-musl`            | 1.1.20 | 6.3.0   |     | N/A   |   ✓    |
| `mips-unknown-linux-gnu`             | 2.23   | 5.3.1   | ✓   | 5.1.0 |   ✓    |
| `mips-unknown-linux-musl`            | 1.1.20 | 6.3.0   | ✓   | 5.1.0 |   ✓    |
| `mips64-unknown-linux-gnuabi64`      | 2.23   | 5.3.1   | ✓   | 5.1.0 |   ✓    |
| `mips64el-unknown-linux-gnuabi64`    | 2.23   | 5.3.1   | ✓   | 5.1.0 |   ✓    |
| `mipsel-unknown-linux-gnu`           | 2.23   | 5.3.1   | ✓   | 5.1.0 |   ✓    |
| `mipsel-unknown-linux-musl`          | 1.1.20 | 6.3.0   | ✓   | 5.1.0 |   ✓    |
| `powerpc-unknown-linux-gnu`          | 2.19   | 4.8.2   | ✓   | 5.1.0 |   ✓    |
| `powerpc64-unknown-linux-gnu`        | 2.31   | 10.2.0  | ✓   | 5.1.0 |   ✓    |
| `powerpc64le-unknown-linux-gnu`      | 2.19   | 4.8.2   | ✓   | 5.1.0 |   ✓    |
| `riscv64gc-unknown-linux-gnu`        | 2.27   | 7.5.0   | ✓   | 5.1.0 |   ✓    |
| `s390x-unknown-linux-gnu`            | 2.23   | 5.3.1   | ✓   | 5.1.0 |        |
| `sparc64-unknown-linux-gnu`          | 2.31   | 10.2.0  | ✓   | 5.1.0 |   ✓    |
| `sparcv9-sun-solaris` [4]            | 2.11   | 5.3.0   | ✓   | N/A   |        |
| `thumbv6m-none-eabi` [5]             | 2.2.0  | 5.3.1   |     | N/A   |        |
| `thumbv7em-none-eabi` [5]            | 2.2.0  | 5.3.1   |     | N/A   |        |
| `thumbv7em-none-eabihf` [5]          | 2.2.0  | 5.3.1   |     | N/A   |        |
| `thumbv7m-none-eabi` [5]             | 2.2.0  | 5.3.1   |     | N/A   |        |
| `wasm32-unknown-emscripten` [6]      | 1.2.2  | 3.1.10  | ✓   | N/A   |   ✓    |
| `x86_64-linux-android` [2]           | 9.0.0  | 4.9     | ✓   | 5.1.0 |   ✓    |
| `x86_64-pc-windows-gnu`              | N/A    | 7.3.0   | ✓   | N/A   |   ✓    |
| `x86_64-sun-solaris` [4]             | 2.11   | 5.3.0   | ✓   | N/A   |        |
| `x86_64-unknown-freebsd` [4]         | 12.1   | 6.4.0   |     | N/A   |   ✓    |
| `x86_64-unknown-dragonfly` [4] [3]   | 4.6.0  | 5.3.0   | ✓   | N/A   |        |
| `x86_64-unknown-linux-gnu`           | 2.15   | 4.6.2   | ✓   | N/A   |   ✓    |
| `x86_64-unknown-linux-musl`          | 1.1.20 | 6.3.0   |     | N/A   |   ✓    |
| `x86_64-unknown-netbsd` [4]          | 7.0    | 5.3.0   | ✓   | N/A   |        |
```

I've removed the offending commit, simplified the commit history (so the commits can be more expressive and remove some commits that were later reverted) and updating the README as we speak.

Alexhuszagh commented:
One more question: I'm assuming the GCC version is for the cross-compiler (GCC if applicable, for Android, clang)? Or is it for the actual GCC version installed for x86_64?

Emilgardis commented:
it's what the cross-compiler would see yes

Alexhuszagh commented:
bors try

Alexhuszagh commented:
Ok the README should be updated with all the current Qemu, libc, and GCC versions. The commit history is simplified and better documented, and all the relevant tests pass. I believe this is ready to merge.

otavio commented:
@Alexhuszagh can you rebase? What else need fixing so this could be merged?

Alexhuszagh commented:
> @Alexhuszagh can you rebase? What else need fixing so this could be merged?

I'll rebase later but this can't be merged as-is due to breaking changes.

Alexhuszagh commented:
Ok this has been rebased, but still should not be approved yet. However, you're free to use these images in your desired workflows using these pre-built images.

otavio commented:
@Alexhuszagh could you open to do items as comment about what is still pending?

Alexhuszagh commented:
We have a few more targets to update (see the images I re-enabled and the new targets I added), so this isn't complete right now.

Alexhuszagh commented:
This actually works locally on my machine, which is very unusual.
https://github.com/cross-rs/cross/runs/6922018754?check_suite_focus=true#step:10:472

Emilgardis commented:
are you using the same cache as CI?

Alexhuszagh commented:
> are you using the same cache as CI?

Ah no, I didn't add the `--from-ci` flag, we might need a fresh rebuild. I should review the weekly update PR.

Alexhuszagh commented:
> ~are you using the same cache as CI?~ no cache on ci, so maybe the question should be, are you being cached?

Um, I'm running from step 1. Trying now with the `--from-ci` flag for `powerpc64le-unknown-linux-gnu`, which I haven't previously built locally. `i686-unknown-linux-gnu` ran locally without issue without a cache, I believe. I can remove all local images, prune, and try again.

Alexhuszagh commented:
bors try --target x86_64-unknown-linux-gnu --target x86_64-unknown-linux-gnu:centos --target aarch64-unknown-linux-gnu --target arm-unknown-linux-gnueabihf --target armv7-unknown-linux-gnueabihf --target thumbv7neon-unknown-linux-gnueabihf --target i686-unknown-linux-gnu --target mipsel-unknown-linux-gnu --target mips64el-unknown-linux-gnuabi64 --target powerpc64le-unknown-linux-gnu --target s390x-unknown-linux-gnu --target i686-pc-windows-gnu

Alexhuszagh commented:
bors try *-gnu*

Alexhuszagh commented:
bors try --target *-gnu*

Alexhuszagh commented:
bors cancel

Alexhuszagh commented:
bors try-
bors try --target *-gnu*

Alexhuszagh commented:
bors try-
bors try

Alexhuszagh commented:
bors try --target x86_64-unknown-dragonfly

Alexhuszagh commented:
bors try --target x86_64-unknown-linux-gnu*

Alexhuszagh commented:
bors try-
bors try --target x86_64-unknown-linux-gnu bors try --target x86_64-unknown-linux-gnu.centos

Alexhuszagh commented:
bors try --target x86_64-unknown-dragonfly

Alexhuszagh commented:
bors try --target x86_64-sun-solaris --target sparcv9-sun-solaris

Alexhuszagh commented:
bors try --target sparcv9-sun-solaris

Emilgardis reviewed (approved):
need to update https://github.com/cross-rs/cross/blob/48d2b7be508f7e8bc3375dc5d2a9ae8da89e62b6/src/docker/shared.rs#L21

otherwise this should be ready!

Alexhuszagh commented:
> need to update
> 
> https://github.com/cross-rs/cross/blob/48d2b7be508f7e8bc3375dc5d2a9ae8da89e62b6/src/docker/shared.rs#L21
> 
> otherwise this should be ready!

We should probably add a script to check this is compatible, and then install it in our hooks. Wouldn't be too hard. This ensures we have the correct base image in the future.

Emilgardis commented:
instead of a hook, I think it makes sense to have it as a test.

Emilgardis commented:
something like this: https://github.com/cross-rs/cross/blob/1efbfbace98ad767155e9a711525192ddab458e7/xtask/src/codegen.rs#L73-L79

Emilgardis commented on CHANGELOG.md:
```diff
@@ -122,6 +122,7 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 - #652 - Allow trying individual targets via bors.
 - #650 - Improve Docker caching.
 - #609 - Switch to Github Actions and GHCR.
+- #591 - Update Ubuntu images to 20.04 LTS.
```
changelog is in the wrong place now

Emilgardis reviewed (approved):
lgtm with changelog fix

Emilgardis reviewed (approved):
bors r+

we're finally in 2020 🎉 🎉

Alexhuszagh commented:
So it seems the `mips-unknown-linux-gnu` failure was caused by #724. The i686 FreeBSD build might be a bit more of an issue...

bors try --target mips-unknown-linux-gnu

Emilgardis commented on docker/Dockerfile.armv7-unknown-linux-gnueabihf:
```diff
@@ -1,4 +1,4 @@
-FROM ubuntu:16.04
+FROM ubuntu:20.04
```
This image, and probably more, reference a `cc` that is not in path: `arm-linux-gnueabihf-gcc` does not exist.

Emilgardis replied:
discussed in matrix

https://matrix.to/#/!ZFqvHUjUqgONNObpSJ:matrix.org/$fNBbYZ_-RWoY5pNjbyMKkLKFedrfqYahsuMZDZr_dhI?via=matrix.org&via=matrix.questionable.software&via=rduce.org

Emilgardis replied:
this is wrong, it does exist

```
$ cargo build-docker-image armv7-unknown-linux-gnueabihf --tag ubuntu20
[..snip]
 => => writing image sha256:fa5d48f3e9c18589f373cdeece4b611c1d5f3dcea31cdc2d25ce0c2d575f4b23                                                                                                                    0.0s
 => => naming to ghcr.io/cross-rs/armv7-unknown-linux-gnueabihf:ubuntu20   
$ docker run --rm -it ghcr.io/cross-rs/armv7-unknown-linux-gnueabihf:ubuntu20
root@df34d3813315:/# arm-linux-gnueabihf-
arm-linux-gnueabihf-addr2line     arm-linux-gnueabihf-dwp           arm-linux-gnueabihf-gcc-ar        arm-linux-gnueabihf-gcov          arm-linux-gnueabihf-gprof         arm-linux-gnueabihf-objdump
arm-linux-gnueabihf-ar            arm-linux-gnueabihf-elfedit       arm-linux-gnueabihf-gcc-ar-9      arm-linux-gnueabihf-gcov-9        arm-linux-gnueabihf-ld            arm-linux-gnueabihf-ranlib
arm-linux-gnueabihf-as            arm-linux-gnueabihf-g++           arm-linux-gnueabihf-gcc-nm        arm-linux-gnueabihf-gcov-dump     arm-linux-gnueabihf-ld.bfd        arm-linux-gnueabihf-readelf
arm-linux-gnueabihf-c++filt       arm-linux-gnueabihf-g++-9         arm-linux-gnueabihf-gcc-nm-9      arm-linux-gnueabihf-gcov-dump-9   arm-linux-gnueabihf-ld.gold       arm-linux-gnueabihf-size
arm-linux-gnueabihf-cpp           arm-linux-gnueabihf-gcc           arm-linux-gnueabihf-gcc-ranlib    arm-linux-gnueabihf-gcov-tool     arm-linux-gnueabihf-nm            arm-linux-gnueabihf-strings
arm-linux-gnueabihf-cpp-9         arm-linux-gnueabihf-gcc-9         arm-linux-gnueabihf-gcc-ranlib-9  arm-linux-gnueabihf-gcov-tool-9   arm-linux-gnueabihf-objcopy       arm-linux-gnueabihf-strip
````

but, installing binutils removes it

```
root@df34d3813315:/# apt-get install --no-install-recommends binutils-arm-linux-gnueabihf:armhf binutils:armhf libclang-10-dev:armhf clang-10:armhf
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages were automatically installed and are no longer required:
  binutils-common binutils-x86-64-linux-gnu cpp cpp-9 cpp-9-arm-linux-gnueabihf cpp-arm-linux-gnueabihf gcc-10-cross-base gcc-9-arm-linux-gnueabihf-base gcc-9-base gcc-9-cross-base lib32gcc-s1 lib32stdc++6
  libasan5 libasan5-armhf-cross libatomic1 libatomic1-armhf-cross libbinutils libbsd0 libc6-dev libc6-i386 libcc1-0 libcrypt-dev libctf-nobfd0 libctf0 libedit2 libgc1c2 libgcc-9-dev libgcc-9-dev-armhf-cross
  libgcc-s1-armhf-cross libgomp1 libgomp1-armhf-cross libisl22 libitm1 libllvm10 liblsan0 libmpc3 libmpfr6 libobjc-9-dev libobjc4 libquadmath0 libstdc++-9-dev libstdc++-9-dev-armhf-cross libstdc++6-armhf-cross
  libtsan0 libubsan1 libubsan1-armhf-cross linux-libc-dev
Use 'apt autoremove' to remove them.
The following additional packages will be installed:
  binutils-common:armhf gcc-9-base:armhf libasan5:armhf libatomic1:armhf libbinutils:armhf libbsd0:armhf libclang-common-10-dev:armhf libclang-cpp10:armhf libclang1-10:armhf libctf-nobfd0:armhf libctf0:armhf
  libedit2:armhf libffi7:armhf libgc1c2:armhf libgcc-9-dev:armhf libgomp1:armhf libllvm10:armhf libobjc-9-dev:armhf libobjc4:armhf libstdc++-9-dev:armhf libstdc++6:armhf libubsan1:armhf
Suggested packages:
  binutils-doc:armhf clang-10-doc:armhf libstdc++-9-doc:armhf
Recommended packages:
  llvm-10-dev:armhf python3:armhf libomp-10-dev:armhf
The following packages will be REMOVED:
  binutils binutils-arm-linux-gnueabihf g++ g++-9 g++-9-arm-linux-gnueabihf g++-arm-linux-gnueabihf gcc gcc-9 gcc-9-arm-linux-gnueabihf gcc-arm-linux-gnueabihf libclang-10-dev libclang-common-10-dev libclang-dev
  libclang1-10 libtool
The following NEW packages will be installed:
  binutils:armhf binutils-arm-linux-gnueabihf:armhf binutils-common:armhf clang-10:armhf gcc-9-base:armhf libasan5:armhf libatomic1:armhf libbinutils:armhf libbsd0:armhf libclang-10-dev:armhf
  libclang-common-10-dev:armhf libclang-cpp10:armhf libclang1-10:armhf libctf-nobfd0:armhf libctf0:armhf libedit2:armhf libffi7:armhf libgc1c2:armhf libgcc-9-dev:armhf libgomp1:armhf libllvm10:armhf
  libobjc-9-dev:armhf libobjc4:armhf libstdc++-9-dev:armhf libstdc++6:armhf libubsan1:armhf
0 upgraded, 26 newly installed, 15 to remove and 12 not upgraded.
Need to get 58.0 MB of archives.
After this operation, 35.7 MB disk space will be freed.
```

this is just a pitfall, installing `clang:armhf` etc is wrong, iirc, it's possible to set some packages as forbidden, could we add all the packages that would uninstall the build tools to such a banlist?

Alexhuszagh replied:
Ok we can use `/etc/apt/preferences.d` with pinning to forbid the installation of these packages:

**cat /etc/apt/preferences.d/clang**

```text
Package: clang-10:armhf
Pin: release *
Pin-Priority: -1
```

And this gives:

```bash
$ apt-get install clang-10:armhf
Reading package lists... Done
Building dependency tree
Reading state information... Done
Package clang-10:armhf is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'clang-10:armhf' has no installation candidate
```

Emilgardis commented:
> So it seems the `mips-unknown-linux-gnu` failure was caused by #724

the cache was restored from [v0-rust-build-1.62.0-x86_64-unknown-linux-gnu-a8314ef7d0ec-f9b18b9557c24afdba54](https://github.com/cross-rs/cross/runs/7268777919?check_suite_focus=true#step:3:102), which is the same cache as used in https://github.com/cross-rs/cross/actions/runs/2642147560

We might need to deploy the remedies we know of, or always first build cross with a cross image that has glibc 2.17. or simply cache per target (unsure if that could cause issues with size though)

Emilgardis commented:
try again?

Alexhuszagh commented:
> try again?

I still have to fix the FreeBSD issue.

Alexhuszagh commented:
bors try --target x86_64-unknown-freebsd

Alexhuszagh commented:
bors try --target x86_64-unknown-freebsd

Alexhuszagh commented:
bors r=Emilgardis

Alexhuszagh commented:
bors try --target mips-unknown-linux-gnu

Alexhuszagh commented:
bors r=Emilgardis
